Fix #595: Fail Fast on Step Nesting

BrunoJuchli opened the pull request:
This fixes ~#595~ #617.

When outside of collection phase (defined by `ScenarioInvoker.InvokeScenarioMethodAsync`), throws an `InvalidOperationException` on read or write access to the collection accessed via `CurrentThread.StepDefinitions`.
Accessing the property itself doesn't throw, but access methods/properties on the collection does.

When the client code is storing a reference to the collection for later use, and that use is after collection phase, using the collection will throw.

### Breaking Changes

It's now not possible anymore to add steps before collection phase has started.
An example that doesn't work anymore:

    public class static Foo
    {
         static Foo()
         {
             "Must start first test with this".x(() => {});
         }
     }

adamralph commented:
@BrunoJuchli thanks very much for raising this, and congratulations for finding a way to do it! 

Essentially you've added a "gate" to adding steps which is opened and closed at the right times in `ScenarioInvoker`. This makes sense. I've boiled down the implementation to somthing simpler in the latest commit. Can you please take a look and see what you think?

adamralph committed:
red: add ScenarioFeature.NestedStep

Co-authored-by: Juchli Bruno Lorenz <bruno.lorenz.juchli@erowa.com>

adamralph committed:
green: implement ScenarioFeature.NestedStep

throw an exception when a step is defined in the wrong place, such as within another step

Co-authored-by: Juchli Bruno Lorenz <bruno.lorenz.juchli@erowa.com>

BrunoJuchli commented:
@adamralph Yours is simpler thus better.
I initially did a similar implementation, then reverted, because I thought it might be too much of a breaking change.

There's one draw back though: Out of order disposal / double disposal will break the logic and won't be easy to debug.

I may be a bit overly sensitive with this one, though :D
When I first implemented this I originally did:

    CurrentThread.BeginCollection()
    try {...}
    finally
    {
        CurrentThread.EndCollection()
    }

The result was, that one of the tests failed @ `CurrentThread.EndCollection`. It seems it was executed outside of the synchronization context. Thus I changed the implementation to:

    var reporter = CurrentThread.BeginCollection();
    try {...}
    finally
    {
        reporter.EndCollection();
     }

Another reason for my `Dispose` sensitivity is that I once read that there's situations where you'd write code that involved`using` statements and what the .net framework would execute would be unexpected: in exceptional situations it'd execute the dispose statement twice. Too bad I can't find the post anymore. I did trust it because it had a good explanation of the inner workings of the language / .Net Framework.

adamralph commented:
@BrunoJuchli I'm not sure I understand the problem. What do you mean by "Out of order disposal / double disposal"? And what are `BeginReporting` and `EndReporting`?

BrunoJuchli commented:
@adamralph 
Re `BeginReporting` and `EndReporting`. Sorry, I got that mixed up. I meant to write `BeginCollection` and `EndCollection` (what is now `using (new StepDefinitionContext()) {...}`).

Out of order disposal:

    var context1 = new StepDefinitionContext();
    var context2 = new StepDefinitionContext();
    context1.Dispose();
    context2.Dispose();

Double disposal:

    var context1 = new StepDefinitionContext();
    context1.Dispose();
    var context2 = new StepDefinitionContext();
    context1.Dispose();
    context2.Dispose();

The current implementation is fine, there's no issue. There's just a slight risk that the future will bring concurrency related problems and the current implementation will not make it exactly easy to pinpoint them (it'll just behave incorrectly and not ever throw an exception). I guess you have more experience with that sort of thing and that if you deem it unnecessary to take extra precautions that it'll be fine.

adamralph commented:
@BrunoJuchli that would be a misuse of the API, so I'm not worried about it. Also, this is all part of the SDK (in the `Xbehave.Sdk` namespace) which I imagine very few, if any, people are using.

That said, I think I will make `StepDefinitionContext` `internal` and instead expose a  method on `CurrentThread` which returns an `IDisposable`. That way the API should be more obvious.

adamralph commented:
@BrunoJuchli can you please take a look at the latest diff and see what you think?

Note that I've marked the commits as co-authored by the two of us.
